src/custom-element.js

```diff
@@ -591,13 +592,145 @@ function createBaseCustomElementClass(win, elementConnectedCallback) {
       );
       return readyPromise.then(() => {
         if (this.V1()) {
-          const builder = getBuilderForDoc(this.getAmpDoc());
-          builder.scheduleAsap(this);
+          const scheduler = getSchedulerForDoc(this.getAmpDoc());
+          scheduler.scheduleAsap(this);
         }
         return this.whenBuilt();
       });
     }
 
+    /**
+     * Mounts the element by calling the `BaseElement.mountCallback` method.
+     *
+     * Can only be called on a upgraded element. May only be called from
+     * scheduler.js.
+     *
+     * @return {!Promise}
+     * @final
+     * @restricted
+     */
+    mountInternal() {
+      if (this.mountPromise_) {
+        return this.mountPromise_;
+      }
+      this.mountAbortController_ = new AbortController();
+      const {signal} = this.mountAbortController_;
+      return (this.mountPromise_ = this.buildInternal()
+        .then(() => {
+          devAssert(this.V1());
+          if (signal.aborted) {
+            // Mounting has been canceled.
+            return;
+          }
+          this.setReadyStateInternal(
+            this.implClass_.load(this)
+              ? ReadyState.LOADING
+              : ReadyState.MOUNTING
+          );
+          return this.impl_.mountCallback(signal);
+        })
+        .then(() => {
+          this.mountAbortController_ = null;
+          if (signal.aborted) {
+            throw cancellation();
+          }
+          this.signals_.signal(CommonSignals.MOUNTED);
+          if (
+            this.implClass_.load(this) &&
+            this.readyState_ !== ReadyState.COMPLETE
+          ) {
+            this.setReadyStateInternal(ReadyState.LOADING);
+          } else {
+            this.setReadyStateInternal(ReadyState.COMPLETE);
+          }
+        })
+        .catch((reason) => {
+          this.mountAbortController_ = null;
+          if (isCancellation(reason)) {
+            this.mountPromise_ = null;
```

Comment:
There's a race here with `unmount` then `mount` before this catch can fire. We'll drop the `mountPromise_` opening up the possibility `mount` being called a second time.

Reply:
`unmount()` itself also resets `mountPromise_` to `null`. And it's also why `isCancellation()` fork is ignored. Do you think the race condition is still possible?

Reply:
Yes, it doesn't matter that `unmount` deletes the prop. I believe this is easily reproduced with just:

```js
// setup a impl.mountCallback counter, not shown
element.mount();
element.unmount();
await element.mount();
assert(counter === 1); // But I think this will be 2.

Reply:
Yep. You're right. I changed this a bit and now I'm tracking `mount->unmount` sequence via `mountAbortController` to avoid race condition. The `counter === 1` is now ensured. PTAL.